base/array.jl

```diff
@@ -1987,3 +1987,162 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
 ```
 """
 symdiff(a, b, rest...) = symdiff(a, symdiff(b, rest...))
+
+## replace/replace! ##
+
+# NOTE: to implement replace! and replace for a new type T, it's enough to define:
+# function _replace!(pred::Callable, new::Callable, A::T, n::Int)
+
+"""
+    replace!(A, old_new::Pair...; [n::Integer])
+
+For each pair `old=>new` in `old_new`, replace all occurrences
+of `old` in collection `A` by `new`.
+If `n` is specified, then replace at most `n` occurrences in total.
+See also [`replace`](@ref replace(A, old_new::Pair...)).
+
+# Examples
+```jldoctest
+julia> replace!([1, 2, 1, 3], 1=>0, 2=>4, n=2)
+4-element Array{Int64,1}:
+ 0
+ 4
+ 1
+ 3
+
+julia> replace!(Set([1, 2, 3]), 1=>0)
+Set([0, 2, 3])
+```
+"""
+replace!(A, old_new::Pair...; n::Integer=typemax(Int)) = _replace!(A, eltype(A), n, old_new...)
+
+_replace!(A, ::Type{K}, n::Integer, old_new::Pair...) where {K} = _replace!(A, n) do x
```

Comment:
This comment failed to be sent yesterday:
> Also, this was necessary to introduce a helper method with `n` as a positional arg instead of a keyword arg, as otherwise forwarding arguments with `n=n` for the keyword was killing performances (in particular when the initial call used the default arg).

I benchmarked `replace` with Benchmarktool, but I need to do it more precisely. In the meantime, I implemented a more direct algorithm for `replace!` taking pairs, and it's faster by about 50-100%. For a base function, maybe it's worth it.

>  Naming helpers differently is fine, but here it wouldn't make any difference AFAIK.

I don't see how it doesn't make a difference (or I misunderstand what you mean): if `_replace!(A, ::Type{K}, n::Integer, old_new::Pair...)` was named `replace!`, this method would not be more discoverable but not part of the API...